{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-05-06T03:03:13.269100Z",
     "start_time": "2022-05-06T03:03:13.262488Z"
    }
   },
   "outputs": [],
   "source": [
    "# convert coverage tracks in bedgraph (bdg) format of eg RNA-seq or ATAC-seq into my circos format"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-08-09T23:21:24.819199Z",
     "start_time": "2022-08-09T23:21:23.647372Z"
    }
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import pybedtools\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-08-09T23:22:05.840968Z",
     "start_time": "2022-08-09T23:22:05.832333Z"
    }
   },
   "outputs": [],
   "source": [
    "def generate_circos_coverage(bdg_file,karyotype_file):\n",
    "    cov = pybedtools.BedTool(bdg_file)\n",
    "    df = pd.DataFrame()\n",
    "    with open(karyotype_file,'r') as f:\n",
    "        for line in f.readlines():\n",
    "            line = line.split(' ')\n",
    "            loc = line[3]+' '+line[4]+' '+line[5]\n",
    "            kary = pybedtools.BedTool(loc, from_string=True)\n",
    "            res = cov.intersect(kary)\n",
    "            del(kary)\n",
    "            try:\n",
    "                res = res.to_dataframe(disable_auto_names=True,header=None)\n",
    "            except pd.errors.EmptyDataError:\n",
    "                continue\n",
    "            res[\"circos_chr\"] = line[2]\n",
    "            df = pd.concat([df,res])\n",
    "    del cov\n",
    "    df.columns = ['original_chr','start','end','score','circos_chr']\n",
    "    df = df[[\"circos_chr\",'start','end','score']]\n",
    "    outfile = os.path.join(os.path.dirname(karyotype_file),os.path.basename(bdg_file)+'.circos.bdg')\n",
    "    df.to_csv(outfile,sep='\\t',header=False,index=False)\n",
    "    return df\n",
    "    \n",
    "def import_bdg_file(bdg_file):\n",
    "    df = pd.read_table(bdg_file,header=None,sep='\\t',\n",
    "                       names=['chr','start','end','score'])\n",
    "    return df\n",
    "def import_karyotype_file(karyotype_file):\n",
    "    df = pd.read_table(karyotype_file,header=None,sep=' ',\n",
    "                       names=['chr','-','n','chrom','start','end','color'],\n",
    "                       usecols=['n','chrom','start','end'])\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-08-09T23:22:32.073680Z",
     "start_time": "2022-08-09T23:22:06.836039Z"
    },
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>circos_chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>31-</td>\n",
       "      <td>56797826</td>\n",
       "      <td>56798000</td>\n",
       "      <td>51</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>31-</td>\n",
       "      <td>56798000</td>\n",
       "      <td>56799000</td>\n",
       "      <td>725</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>31-</td>\n",
       "      <td>56799000</td>\n",
       "      <td>56800000</td>\n",
       "      <td>338</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>31-</td>\n",
       "      <td>56800000</td>\n",
       "      <td>56801000</td>\n",
       "      <td>137</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>31-</td>\n",
       "      <td>56801000</td>\n",
       "      <td>56802000</td>\n",
       "      <td>417</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>197</th>\n",
       "      <td>28-29+</td>\n",
       "      <td>56790000</td>\n",
       "      <td>56791000</td>\n",
       "      <td>347</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>198</th>\n",
       "      <td>28-29+</td>\n",
       "      <td>56791000</td>\n",
       "      <td>56792000</td>\n",
       "      <td>437</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>199</th>\n",
       "      <td>28-29+</td>\n",
       "      <td>56792000</td>\n",
       "      <td>56793000</td>\n",
       "      <td>239</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>200</th>\n",
       "      <td>28-29+</td>\n",
       "      <td>56793000</td>\n",
       "      <td>56794000</td>\n",
       "      <td>173</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>201</th>\n",
       "      <td>28-29+</td>\n",
       "      <td>56794000</td>\n",
       "      <td>56794203</td>\n",
       "      <td>35</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2566 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    circos_chr     start       end  score\n",
       "0          31-  56797826  56798000     51\n",
       "1          31-  56798000  56799000    725\n",
       "2          31-  56799000  56800000    338\n",
       "3          31-  56800000  56801000    137\n",
       "4          31-  56801000  56802000    417\n",
       "..         ...       ...       ...    ...\n",
       "197     28-29+  56790000  56791000    347\n",
       "198     28-29+  56791000  56792000    437\n",
       "199     28-29+  56792000  56793000    239\n",
       "200     28-29+  56793000  56794000    173\n",
       "201     28-29+  56794000  56794203     35\n",
       "\n",
       "[2566 rows x 4 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "karyotype = '/mnt/c/Users/ochapman/Documents/Mesirov/medullo_ecDNA/src/circos/D458/D458_ecDNA.karyotype'\n",
    "import_karyotype_file(karyotype)\n",
    "bdg = '/mnt/c/Users/ochapman/Documents/Mesirov/medullo_ecDNA/src/circos/D458/atac/D458.hg38.bdg'\n",
    "import_bdg_file(bdg)\n",
    "df = generate_circos_coverage(bdg,karyotype)\n",
    "df\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# D458 CRISPRi proliferation screen\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-05-06T03:29:15.992646Z",
     "start_time": "2022-05-06T03:29:15.987831Z"
    }
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import pyranges as pr\n",
    "import os\n",
    "import io"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-05-06T03:39:55.723254Z",
     "start_time": "2022-05-06T03:39:55.712601Z"
    }
   },
   "outputs": [],
   "source": [
    "def generate_circos_coverage(bdg_file,karyotype_file,outfile):\n",
    "    cov = import_bdg_file(bdg_file)\n",
    "    df = pd.DataFrame()\n",
    "    with open(karyotype_file,'r') as f:\n",
    "        for line in f.readlines():\n",
    "            kary = io.StringIO(line)\n",
    "            kary = import_karyotype_file(kary)\n",
    "            res = cov.intersect(kary)\n",
    "            del(kary)\n",
    "            res = res.as_df()\n",
    "            res[\"circos_chr\"] = line.strip().split()[2]\n",
    "            df = pd.concat([df,res])\n",
    "    del cov\n",
    "    df.columns = ['original_chr','start','end','score','circos_chr']\n",
    "    df = df[[\"circos_chr\",'start','end','score']]\n",
    "    df.to_csv(outfile,sep='\\t',header=False,index=False)\n",
    "    return df\n",
    "    \n",
    "def import_bdg_file(bdg_file):\n",
    "    df = pd.read_table(bdg_file,header=None,sep='\\t',\n",
    "                       names=['Chromosome','Start','End','Score'])\n",
    "    return pr.PyRanges(df)\n",
    "\n",
    "    return df\n",
    "def import_karyotype_file(karyotype_file):\n",
    "    df = pd.read_table(karyotype_file,header=None,sep=' ',\n",
    "                       names=['chr','-','n','Chromosome','Start','End','color'],\n",
    "                       usecols=['n','Chromosome','Start','End'])\n",
    "    df = pr.PyRanges(df)\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-05-06T03:44:22.631340Z",
     "start_time": "2022-05-06T03:44:21.619404Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>circos_chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2</td>\n",
       "      <td>56797991</td>\n",
       "      <td>56797991</td>\n",
       "      <td>0.005159</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>56798011</td>\n",
       "      <td>56798011</td>\n",
       "      <td>0.005159</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>56798031</td>\n",
       "      <td>56798031</td>\n",
       "      <td>0.005159</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2</td>\n",
       "      <td>56798051</td>\n",
       "      <td>56798051</td>\n",
       "      <td>0.005159</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2</td>\n",
       "      <td>56798071</td>\n",
       "      <td>56798071</td>\n",
       "      <td>0.005159</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1405</th>\n",
       "      <td>1</td>\n",
       "      <td>56791231</td>\n",
       "      <td>56791231</td>\n",
       "      <td>0.003505</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1406</th>\n",
       "      <td>1</td>\n",
       "      <td>56791251</td>\n",
       "      <td>56791251</td>\n",
       "      <td>0.003505</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1407</th>\n",
       "      <td>1</td>\n",
       "      <td>56791271</td>\n",
       "      <td>56791271</td>\n",
       "      <td>0.003505</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1408</th>\n",
       "      <td>1</td>\n",
       "      <td>56791291</td>\n",
       "      <td>56791291</td>\n",
       "      <td>0.003505</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1409</th>\n",
       "      <td>1</td>\n",
       "      <td>56793011</td>\n",
       "      <td>56793011</td>\n",
       "      <td>-0.002762</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>18107 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     circos_chr     start       end     score\n",
       "0             2  56797991  56797991  0.005159\n",
       "1             2  56798011  56798011  0.005159\n",
       "2             2  56798031  56798031  0.005159\n",
       "3             2  56798051  56798051  0.005159\n",
       "4             2  56798071  56798071  0.005159\n",
       "...         ...       ...       ...       ...\n",
       "1405          1  56791231  56791231  0.003505\n",
       "1406          1  56791251  56791251  0.003505\n",
       "1407          1  56791271  56791271  0.003505\n",
       "1408          1  56791291  56791291  0.003505\n",
       "1409          1  56793011  56793011 -0.002762\n",
       "\n",
       "[18107 rows x 4 columns]"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# D458 CRISPRi proliferation screen\n",
    "karyotype =  'D458/D458_ecDNA.karyotype'\n",
    "bdg = '/mnt/c/Users/ochapman/Documents/Mesirov/medullo_ecDNA/crispri/2022-03-31_crispr-surf/D458/deconvolved_scores.bedgraph'\n",
    "#import_bdg_file(bdg)\n",
    "#import_karyotype_file(kary)\n",
    "generate_circos_coverage(bdg,karyotype,'D458/crispri/D458_crispri.hg38.circos.bdg')\n",
    "\n",
    "bdg = '/mnt/c/Users/ochapman/Documents/Mesirov/medullo_ecDNA/crispri/2022-03-31_crispr-surf/D283/deconvolved_scores.bedgraph'\n",
    "generate_circos_coverage(bdg,karyotype,'D458/crispri/D283_crispri.hg38.circos.bdg')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-05-06T03:34:31.923057Z",
     "start_time": "2022-05-06T03:34:31.853132Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr8</td>\n",
       "      <td>126957191</td>\n",
       "      <td>126957191</td>\n",
       "      <td>-0.005884</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr8</td>\n",
       "      <td>126957211</td>\n",
       "      <td>126957211</td>\n",
       "      <td>-0.005884</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr8</td>\n",
       "      <td>126957231</td>\n",
       "      <td>126957231</td>\n",
       "      <td>-0.005884</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr8</td>\n",
       "      <td>126957251</td>\n",
       "      <td>126957251</td>\n",
       "      <td>-0.005884</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr8</td>\n",
       "      <td>126957271</td>\n",
       "      <td>126957271</td>\n",
       "      <td>-0.005884</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14930</th>\n",
       "      <td>chr17</td>\n",
       "      <td>7484991</td>\n",
       "      <td>7484991</td>\n",
       "      <td>-0.003621</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14931</th>\n",
       "      <td>chr17</td>\n",
       "      <td>7485011</td>\n",
       "      <td>7485011</td>\n",
       "      <td>-0.003621</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14932</th>\n",
       "      <td>chr17</td>\n",
       "      <td>7485031</td>\n",
       "      <td>7485031</td>\n",
       "      <td>-0.003621</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14933</th>\n",
       "      <td>chr17</td>\n",
       "      <td>7485051</td>\n",
       "      <td>7485051</td>\n",
       "      <td>-0.003621</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14934</th>\n",
       "      <td>chr17</td>\n",
       "      <td>7485071</td>\n",
       "      <td>7485071</td>\n",
       "      <td>-0.003621</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>14935 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+-------------------+\n",
       "| Chromosome   | Start     | End       | Score             |\n",
       "| (category)   | (int32)   | (int32)   | (float64)         |\n",
       "|--------------+-----------+-----------+-------------------|\n",
       "| chr8         | 126957191 | 126957191 | -0.00588417019407 |\n",
       "| chr8         | 126957211 | 126957211 | -0.00588417021866 |\n",
       "| chr8         | 126957231 | 126957231 | -0.00588417024674 |\n",
       "| chr8         | 126957251 | 126957251 | -0.00588417027846 |\n",
       "| ...          | ...       | ...       | ...               |\n",
       "| chr17        | 7485011   | 7485011   | -0.00362106719439 |\n",
       "| chr17        | 7485031   | 7485031   | -0.00362106722799 |\n",
       "| chr17        | 7485051   | 7485051   | -0.00362106725032 |\n",
       "| chr17        | 7485071   | 7485071   | -0.00362106726603 |\n",
       "+--------------+-----------+-----------+-------------------+\n",
       "Unstranded PyRanges object has 14,935 rows and 4 columns from 3 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome."
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import_bdg_file(bdg)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:pybedtools]",
   "language": "python",
   "name": "conda-env-pybedtools-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
